From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sky Debreuil <skydebreuil@google.com>
Date: Thu, 15 Oct 2026 13:46:24 -0700
Subject: [M156] [a11y] Mark image dirty directly when removing image map

Original change's description:
> [a11y] Mark image dirty directly when removing image map
>
> When removing an HTMLMapElement during immediate tree updates, mark the
> associated image AXObject's children dirty directly instead of calling
> NotifyParentChildrenChanged mid-traversal.
>
> Change-Id: Ia25df628d938a6bd48d17de46c9e716de847eefd
> Bug: 570093145
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8526087
> Reviewed-by: Lucas Radaelli <lucasradaelli@google.com>
> Auto-Submit: Sky Debreuil <skydebreuil@google.com>
> Reviewed-by: David Tseng <dtseng@chromium.org>
> Commit-Queue: Sky Debreuil <skydebreuil@google.com>
> Commit-Queue: David Tseng <dtseng@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1715691}

(cherry picked from commit 83ce9f18d70490edb5d154ff04471f9c9fdde346)

Bug: 574033011,570093145
Change-Id: Ia25df628d938a6bd48d17de46c9e716de847eefd
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8589140
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/8078@{#1246}
Cr-Branched-From: c50edc4214d81e4f944c3415d55104a49e46d214-refs/heads/main@{#1706539}

diff --git a/third_party/blink/renderer/modules/accessibility/ax_object_cache_impl.cc b/third_party/blink/renderer/modules/accessibility/ax_object_cache_impl.cc
index a52f243471f65e5e38c3d91134054aacc90a3f31..90328d6b274353584d7f3d184a750bdab7118049 100644
--- a/third_party/blink/renderer/modules/accessibility/ax_object_cache_impl.cc
+++ b/third_party/blink/renderer/modules/accessibility/ax_object_cache_impl.cc
@@ -2181,7 +2181,14 @@ void AXObjectCacheImpl::RemoveSubtreeInternal(
       // **important**: this call must come before the node traversal remove
       // below since that could remove a child which would cause it to not point
       // to its image parent, making it impossible to notify the parent.
-      NotifyParentChildrenChanged(image_ax_object);
+      if (lifecycle_.GetState() ==
+          AXObjectCacheLifecycle::kProcessDeferredUpdates) {
+        if (AXObject* ancestor = InvalidateChildren(image_ax_object)) {
+          ancestor->ChildrenChangedWithCleanLayout();
+        }
+      } else {
+        NotifyParentChildrenChanged(image_ax_object);
+      }
     }
   }
 
